spec/requests/api_controller_idempotent_action_spec.rb

```diff
@@ -0,0 +1,119 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'digest'
+
+RSpec.describe ApiController, type: :request do
+  include_context 'with undefine mock controller'
+
+  let(:idempotency_key1) { '11111111-1111-1111-1111-111111111111' }
+  let(:idempotency_key2) { '22222222-2222-2222-2222-222222222222' }
+  let(:params1) { { data: 'one' } }
+  let(:params2) { { data: 'two' } }
+  let(:response_json) { JSON.parse(response.body) }
+
+  let(:create_mock_controller) do
+    class MockController < ApiController
+      include Idempotentable
+
+      def authentication_enabled?
+        false # NB: disable authentication to simplify tests (it is tested elsewhere)
+      end
+
+      def custom
+        render json: { hello: request.params[:data] }, status: :ok
+      end
```

Comment:
Following the approach described previously with `MockController` included in the code base but only routed to in test mode, these specs could potentially use a different endpoint without authentication.